### PR TITLE
There seems a bug in the cosine merge layer

### PR DESCRIPTION
**Pull request by largelymfs**
It seems that when combining two tensor V1 and V2, the cosine merge output is not dot(V1, V2) / sqrt(dot(V1,V1) * dot(V2, V2)). Suppose V1 and V2 is (batch_size, dim). The result should be (batch_size, 1) while dot(V1, V2) / sqrt(dot(V1,V1) * dot(V2, V2)) is (batch_size, batch_size). Thus, I change it to the tensordot with parameter `self.dotaxes`.

I was wondering is that reasonable?

### CODE REVIEW

File: keras/layers/core.py [hunk 1]
```diff
@@ -418,9 +418,11 @@ def get_output(self, train=False):
         elif self.mode == 'cos':
             l1 = self.layers[0].get_output(train)
             l2 = self.layers[1].get_output(train)
-            output, _ = theano.scan(lambda v1, v2: T.dot(v1, v2) / T.sqrt(T.dot(v1, v1) * T.dot(v2, v2)),
-                                    sequences=[l1, l2],
-                                    outputs_info=None)
+            # output, _ = theano.scan(lambda v1, v2: T.batched_tensordot(v1, v2, 1) / T.sqrt(T.batched_tensordot(v1, v1, 1) * T.batched_tensordot(v2, v2, 1)),
+            #                         sequences=[l1, l2],
+            #                         outputs_info=None)
```
Comment (fchollet) [inline keras/layers/core.py]: What's up with the commented-out code?

**Comment by fchollet** [comment]
I believe you're right.

File: keras/layers/core.py (hunk 1 above)
Comment (largelymfs) [inline keras/layers/core.py]: I believe the code is logic reasonable but actually it causes some errors in compilation...

**Comment by fchollet** [comment]
I think this has been fixed in a recent PR. That part of the code is better tested now.
